be/src/exec/pipeline/exchange/mcast_local_exchange.cpp

```diff
@@ -0,0 +1,206 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021 StarRocks Limited.
+
+#include "exec/pipeline/exchange/mcast_local_exchange.h"
+
+#include "util/logging.h"
+
+namespace starrocks {
+namespace pipeline {
+
+MultiCastLocalExchanger::MultiCastLocalExchanger(size_t consumer_number)
+        : _mutex(),
+          _consumer_number(consumer_number),
+          _current_acc_chunk_size(0),
+          _progress(consumer_number),
+          _opened_source_opcount(consumer_number) {
+    Cell* dummy = new Cell();
+    _head = dummy;
+    _tail = dummy;
+    for (size_t i = 0; i < consumer_number; i++) {
+        _progress[i] = _tail;
+        _opened_source_opcount[i] = 0;
+    }
+}
+
+MultiCastLocalExchanger::~MultiCastLocalExchanger() {
+    while (_head != nullptr) {
+        auto* t = _head->next;
+        delete _head;
+        _head = t;
+    }
+}
+
+bool MultiCastLocalExchanger::can_push_chunk() const {
+    std::unique_lock l(_mutex);
+    if ((_current_acc_chunk_size - _fast_acc_chunk_size) > config::vector_chunk_size) {
+        return false;
+    }
+    return true;
+}
+
+Status MultiCastLocalExchanger::push_chunk(const vectorized::ChunkPtr& chunk, int32_t sink_driver_sequence) {
+    if (chunk->num_rows() == 0) return Status::OK();
+
+    std::unique_lock l(_mutex);
+
+    int32_t used_count = (_consumer_number - _opened_source_number);
```

Comment:
Why `used_count` is `(_consumer_number - _opened_source_number);`, not 0?

Reply:
because at that time, some consumers may have exited already. And we think those consumers have already used that chunk.

Reply:
Maybe your `_opened_source_number ` is `_closed_source_number`?

Reply:
Yes. But I thikn opened_source_number is more natural. It's initialized to be zero, and ++ when called `open_source_operator`.